### PR TITLE
Add initial support for rjsf in pipeline properties

### LINKED ISSUE #2759: allow setting component inputs from files (KFP & Airflow)
Currently, we only provide two methods to specify component inputs:

1. Plain Strings (no newlines)
2. From the output of a parent node

We can provide a third option that selects a local file as the input. During pipeline run/export we would pass the content of this file as a string to the input. 

For example, if a component takes an SQL query as an input, it would be amazing to just reference a local `query.sql` file, as right now newlines can't even be input:

Screen Shot 2022-06-02 at 16 05 47

---

Considerations:
- KFP generates Argo-Workflow YAML, and there are limits on how big of a YAML Kubenrentes will accept, so we might want to  fail if the input file would cause the YAML to be too big (but I am not sure on the exact max value).

### PR DESCRIPTION
**Pull request by marthacryan**
Corresponding to elyra-ai/pipeline-editor#197. 

Fixes #2759 

Todo:
- [x] basic pipeline properties support
- [x] basic generic node properties support
- [x] basic "canvas properties" support
- [x] finish airflow support
- [x] link in custom components
- [x] link in validation
- [x] add in required fields
- [x] add inputpath as oneof
- [x] frontend: pipeline migration
  - [x] backend: double check that relevant test pipelines are migrated properly (once migration is finalized)
- [ ] go through code to make sure all utilities and edits etc. are all corresponding to the updated schema

Testing Follow Up TODOs:
- [x] section headers: description 
- [x] backend validation requires updates (pending some frontend changes)
- [x] pipeline JSON occasionally gets corrupted (e.g. one node's properties/values find their way into another node's `component_parameters` stanza) when connected 2 nodes (no longer seeing; uncheck if not the case)
- [x] frontend: properties don't always save if you navigate away or sometimes disappear after connecting 2 nodes - possibly related to above task
- [x] frontend: schema shows up in pipeline JSON
- [x] frontend: headers appear in pipeline properties `app_data.properties.pipeline_defaults` stanza
- [x] frontend: headers appear in node properties `component_parameters` stanzas
  - [x] backend: code can be updated (`validation.py:493-500`) after this change is made
- [x] frontend: properties not listed as required in the schema show as required in properties panel
- [x] frontend: change look of some items (see [comment](https://github.com/elyra-ai/elyra/pull/2780#issuecomment-1226803610))
- [x] frontend: pipeline editor canvas goes blank occasionally when editing array values (haven't been able to reproduce recently)
- [x] `inputpath` structure in pipeline JSON needs fixed
- [x] drag and drop of notebooks/scripts
- [x] pipeline properties - placeholders for list-valued properties
- ~[ ] user-friendly key-value inputs (env variables etc all)~ - pushing to next release
- [x] pipeline default property propagation (including display in node properties)
- [x] pipeline defaults: section headers (node defaults, generic node defaults)
- [x] custom components: component description missing
- [x] generic components: [node] label description missing
- [x] certain string property validation (frontend) erroneously require objects after connecting to another node
- [x] properties with empty strings as defaults erroneously pass frontend validation when first dragging component into pipeline
- [x] `component_source` parameter should be moved out of `component_parameters` stanza of pipeline JSON (some frontend work required in order to still display the property in properties panel)
- [x] frontend validation fails for node runtime image even if a pipeline default is supplied 
- [x] frontend: properties keep the value when changing input type (e.g. when changing from raw value to file input type, the raw value default is still shown in the value field)
- [x] frontend: airflow component tool tips not displaying all properties

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by akchinSTC** [comment]
@ajbozarth - Do we think this will make the 3.11 milestone this friday?

**Comment by ptitzler** [comment]
Not specific to RJSF migration but since the PR already modifies `canvas_properties_template.jinja2`, can we add descriptions for the [`inputs`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L22) and [`outputs`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L34) headers like we already have for [`additional properties`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L49)?

image

**Comment by akchinSTC** [comment]
When using the papermill operator, the area for the `notebook` parameter in the component properties side panel isnot rendering, but is a required field. 
image

**Comment by kiersten-stokes** [comment]
> When using the papermill operator, the area for the `notebook` parameter in the component properties side panel isnot rendering, but is a required field. 

I think I found a solution to this! Should be working now

**Comment by ptitzler** [comment]
Re-assigned to 3.12 milestone since the enhancements that users would benefit from (e.g. https://github.com/elyra-ai/elyra/issues/2738, https://github.com/elyra-ai/elyra/issues/2750) won't be ready before cut-off according to https://github.com/elyra-ai/elyra/issues/2738#issuecomment-1219676250

**Comment by ptitzler** [comment]
A few things I didn't see listed in the todo list in the official list.

### Pipeline properties

 - Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
 - Sections are not clearly separated (compare this to the layout used for node properties (shown below), which is somewhat better due to indentation). 
image

   Layout for node properties
   image

Since descriptions (for headings/properties) are now only displayed as tool tips, they are much harder to read if they are longer than a few words. In many cases a new user will likely require additional information, which we make available in the documentation. How can we make links available? Could the (?) icons perhaps serve two purposes: as a visual indication that there is info available (via tool tip) and as an HREF to a URL that would open up in a new browser window if one clicks on the icon.

**Comment by ajbozarth** [comment]
@kiersten-stokes I pushed the fix to migration.

I also updated some of the snapshot files, one of them will need to be regenerated again after we fix the issue with persisting the schema.

As for the complex pipeline test, that one uses accessors that no longer exist to fill out the form, I tried to update it to use the new accessors, but there are none in the new form. @marthacryan

**Comment by marthacryan** [comment]
> * Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
Fixed

> * Sections are not clearly separated (compare this to the layout used for node properties (shown below), which is somewhat better due to indentation).
Fixed

> Since descriptions (for headings/properties) are now only displayed as tool tips, they are much harder to read if they are longer than a few words. In many cases a new user will likely require additional information, which we make available in the documentation. How can we make links available? Could the (?) icons perhaps serve two purposes: as a visual indication that there is info available (via tool tip) and as an HREF to a URL that would open up in a new browser window if one clicks on the icon.
I can add this as a todo, I'd need to discuss this a little further

**Comment by ptitzler** [comment]
> > * Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
> 
> Fixed

Heading font size is still 12px vs text font size of 16px

image

---

[generic node] adjacent buttons should not touch

image

---

tool tips for custom components aren't displayed properly

image


---

combination of several issues if one doesn't enter a text value in list-valued input:
 - set this as pipeline default
    
   image

 - non-existent value rendered  in node properties as pipeline default
    image

  - stored as `null` in pipeline file
    
    image

**Comment by ptitzler** [comment]
@kiersten-stokes is this ready for testing or still in wip? https://github.com/elyra-ai/elyra/pull/2767 indicates that this PR will take care of it

image

**Comment by kiersten-stokes** [comment]
> @kiersten-stokes is this ready for testing or still in wip? #2767 indicates that this PR will take care of it

Yes, this is ready to test! I saw your comment on #2767 as well and added the relevant logic to the `describe` command.

**Comment by marthacryan** [comment]
> Heading font size is still 12px vs text font size of 16px

I think I hadn't updated the link with that fix - it looks right on my end but let me know if you're still seeing it on a more recent build. 

> [generic node] adjacent buttons should not touch

Fixed

> tool tips for custom components aren't displayed properly

Fixed

> * non-existent value rendered  in node properties as pipeline default
>   image

Fixed but I need to go in and prevent empty array values from being saved to the file.

**Comment by ptitzler** [comment]
`make install` in a clean environment fails due to linting issues:

```
yarn eslint --max-warnings=0
yarn run v1.22.4
$ eslint . --fix --ignore-path .gitignore --ext .ts,.tsx,.js --max-warnings=0

/Users/patti/elyra_dev2/martha-rjsf/elyra/packages/pipeline-editor/src/PipelineEditorWidget.tsx
  306:28  warning  Missing return type on function                                                                                        @typescript-eslint/explicit-function-return-type
  338:6   warning  React Hook useCallback has a missing dependency: 'removeNullValues'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 2 problems (0 errors, 2 warnings)

ESLint found too many warnings (maximum: 0).
error Command failed with exit code 1.
```

**Comment by ptitzler** [comment]
@kiersten-stokes: overall file input looks good but I managed to break this scenario:
 - choose an input property that has a default value defined (e.g. for bashoperator use `output_encoding`, which is set to `utf-8`)
 - change the selection from `... a string value` to `... select a file`
 - the default value is now interpreted as a file name, which it isn't

**Comment by ptitzler** [comment]
> > tool tips for custom components aren't displayed properly
> 
> Fixed

@marthacryan now no properties are displayed
image

main branch output:

image

**Comment by ptitzler** [comment]
It appears that we are now persisting a property named `component_source_header` in the `.pipeline` file. This property is only used as a visualization aid ("section header") in the UI and shouldn't be persisted:

From https://github.com/marthacryan/elyra/blob/pipeline-rjsf/elyra/templates/components/canvas_properties_template.jinja2#L123-L129
```    
"component_source_header": {
      "type": "null",
      "title": "Component Source",
      "uihints": {
        "ui:field": "header"
      }
    },
```

File: elyra/tests/pipeline/kfp/test_processor_kfp.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 from elyra.metadata.metadata import Metadata
 from elyra.pipeline.catalog_connector import FilesystemComponentCatalogConnector
 from elyra.pipeline.catalog_connector import UrlComponentCatalogConnector
-from elyra.pipeline.component import Component
+from elyra.pipeline.component import Component, ComponentParameter
```
Comment (ptitzler) [inline elyra/tests/pipeline/kfp/test_processor_kfp.py:28]:
```suggestion
from elyra.pipeline.component import Component
from elyra.pipeline.component import ComponentParameter
```

**Comment by kiersten-stokes** [comment]
> choose an input property that has a default value defined (e.g. for bashoperator use output_encoding, which is set to utf-8)
change the selection from ... a string value to ... select a file
the default value is now interpreted as a file name, which it isn't

@ptitzler yeah, just checked and I see this too. I think the fix will be in the frontend, but I'll double check at the meeting this AM!

**Comment by marthacryan** [comment]
@ptitzler Thanks for checking it out again! I fixed the lint and the extra null value. I'm working on the airflow components tooltips issue

**Comment by ptitzler** [comment]
@ajbozarth migration of the [noaa tutorial pipeline](https://github.com/elyra-ai/examples/tree/main/pipelines/dax_noaa_weather_data) fails:

image

**Comment by ptitzler** [comment]
The validation message is invalid if one does not specify a file as string input:

image

Note the value of `value`, which is not what the user provided as input. The expected value would be null (undefined).

**Comment by ptitzler** [comment]
Tooltip for custom nodes in Airflow are now looking good ...
image

For property values that are provided via input files (`URL` in node `download file` in the example below), can we prepend the file name with a string that indicates that we are displaying a reference, not the value? The `count rows` node is an example of a property that is also defined via reference, specifically a reference to the output of an upstream node:

**Comment by kiersten-stokes** [comment]
> The validation message is invalid if one does not specify a file as string input:

Good point! I've fixed validation to allow an empty string for file input types for non-required properties. From here, the processor behaves the same way as an empty string for raw input as well.

**Comment by ptitzler** [comment]
> > The validation message is invalid if one does not specify a file as string input:
> 
> Good point! I've fixed validation to allow an empty string for file input types for non-required properties. From here, the processor behaves the same way as an empty string for raw input as well.

Thank you! Looking good. It appears to preserve default values, which I would expect to happen since the user didn't provide a property value.

**Review by kiersten-stokes — APPROVED** [review]

**Comment by kevin-bates** [comment]
Congratulations (and thank you :heart:) to everyone that worked on this - it's one for the ages! :tada:

**Comment by thesuperzapper** [comment]
I would also love to thank everyone for their amazing effort!!!

❤️

**Pull request merged**